### PR TITLE
[4.0] More flexible and null value aware datetime range filter in articles model

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for Issue # .

### Summary of Changes

Follow up to PR #26295 , which introduced real NULL values for datetime columns in com_content's database table.

Extends the range filter for datetimes in the articles model (frontend) so that

1. One of the parameters start and end of the range may be ommitted.
2. The date to be filtered might be a real NULL value. The filter query will handle that.

### Testing Instructions

Code review, or use that datetime filtering of the frontend articles model somewhere.

### Expected result

See summary of changes.

### Actual result

None of that.

### Documentation Changes Required

None as far as I can see.

### CODE REVIEW

**Review by wilsonge — APPROVED** [review]

**Comment by richard67** [comment]
@wilsonge Does it need to write testing instructions? Those are the only reason why it is still draft/WiP. I can write tomorrow if necessary. Or would "code review" be enough?

File: components/com_content/Model/ArticlesModel.php [hunk 1]
```diff
@@ -480,18 +480,35 @@ function ($data) use ($db)
...
-				$startDateRange = $db->quote($this->getState('filter.start_date_range'));
-				$endDateRange   = $db->quote($this->getState('filter.end_date_range'));
-				$query->where(
-					'(' . $dateField . ' >= ' . $startDateRange . ' AND ' . $dateField .
-					' <= ' . $endDateRange . ')'
-				);
+				$startDateRange = $db->quote($this->getState('filter.start_date_range', ''));
+				$endDateRange   = $db->quote($this->getState('filter.end_date_range', ''));
+
+				$startDateRangeUsed = !empty($startDateRange);
```
Comment (SharkyKZ) [inline components/com_content/Model/ArticlesModel.php]: These will never be empty because they're quoted.

Comment (richard67) [reply components/com_content/Model/ArticlesModel.php]: @SharkyKZ I don't fully understand. Default parameter which I've added is an empty string. Doesn't that result in true when checking for empty? Or dou you mean they are quoted in addition elsewhere so we have `''''`?  What else could I provide as default for the `$this->getState` call? Null, and then use `is_null`or `=== null`? Please advise.

Comment (SharkyKZ) [reply components/com_content/Model/ArticlesModel.php]: `$db->quote()` has to be applied after comparisons.

The query is already glued with `AND` so I don't think we need to explicitly wrap the statements in `()` here. Something like this should work:

```
$startDateRange = $this->getState('filter.start_date_range', '');
$endDateRange   = $this->getState('filter.end_date_range', '');

if ($startDateRange || $endDateRange)
{
	$query->where($dateField . ' IS NOT NULL');

	if ($startDateRange)
	{
		$query->where($dateField . ' >= ' . $db->quote($startDateRange));
	}

	if ($endDateRange)
	{
		$query->where($dateField . ' <= ' . $db->quote($endDateRange));
	}
}
```

Comment (richard67) [reply components/com_content/Model/ArticlesModel.php]: @SharkyKZ Thanks, will implement tonight after work, German time.

Comment (SharkyKZ) [reply components/com_content/Model/ArticlesModel.php]: If you want to keep `()`, `extendWhere()` could be used instead:

```
if ($startDateRange || $endDateRange)
{
	$where = [$dateField . ' IS NOT NULL'];

	if ($startDateRange)
	{
		$where[] = $dateField . ' >= ' . $db->quote($startDateRange);
	}

	if ($endDateRange)
	{
		$where[] = $dateField . ' <= ' . $db->quote($endDateRange);
	}

	$query->extendWhere('AND', $where, 'AND');
}
```

Comment (richard67) [reply components/com_content/Model/ArticlesModel.php]: OMG, looking on it now, I don't understand how I could code it that way. I should have seen that the dates are already quoted. Will fix later. (Blush)

Comment (richard67) [reply components/com_content/Model/ArticlesModel.php]: Done.

**Comment by richard67** [comment]
@SharkyKZ Done. Thanks for the elegant solution you've proposed. I did it that way.

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**